### PR TITLE
New style where function

### PR DESCRIPTION
**Pull request by kmaehashi**
This PR changes `F.where` function to new style.

### CODE REVIEW

File: chainer/functions/array/where.py [hunk 1]
```diff
@@ -26,11 +26,13 @@ def forward(self, inputs):
...
-        xp = cuda.get_array_module(inputs[0])
-        condition = inputs[0]
-        gx = xp.where(condition, grads[0], 0)
-        gy = xp.where(condition, 0, grads[0])
+    def backward(self, indexes, grad_outputs):
+        condition = self.inputs[0]
+        xp = cuda.get_array_module(condition.data)
+        g, = grad_outputs
+        zeros = xp.zeros(g.shape, dtype=g.dtype)
+        gx, = Where().apply((condition.data, g, zeros))
```
Comment (beam2d) [inline chainer/functions/array/where.py]: Could you skip the computation of `gx`/`gy` if not needed? Whether or not they are needed can be determined by `indexes`.

File: chainer/functions/array/where.py [hunk 2]
```diff
@@ -26,11 +26,13 @@ def forward(self, inputs):
...
         return xp.where(condition, x, y),
 
-    def backward(self, inputs, grads):
-        xp = cuda.get_array_module(inputs[0])
-        condition = inputs[0]
-        gx = xp.where(condition, grads[0], 0)
-        gy = xp.where(condition, 0, grads[0])
+    def backward(self, indexes, grad_outputs):
+        condition = self.inputs[0]
+        xp = cuda.get_array_module(condition.data)
```
Comment (beam2d) [inline chainer/functions/array/where.py]: You have to call `self.retain_inputs` if you need input data, and use `get_retained_inputs` to obtain the input variable (see the implementation of `linear` for example).

Comment (kmaehashi) [reply chainer/functions/array/where.py]: Fixed.
Some other functions are directly using self.input and they must be fixed too I think?
e.g., https://github.com/kmaehashi/chainer/blob/492c9bb61fd44457c5e81df5c87120807110e7b8/chainer/functions/array/permutate.py#L79

File: chainer/functions/array/where.py (hunk 1 above)
Comment (kmaehashi) [reply chainer/functions/array/where.py]: Fixed.

**Comment by beam2d** [comment]
Jenkins, test this please

**Comment by beam2d** [comment]
LGTM

**Pull request merged**